### DISCUSSION TITLE
Understanding Discrepancies in Pydantic Serialization and Deserialization with Nested Lists and Tuples

### DISCUSSION THREAD

**Discussion by jhb123** [pydantic/pydantic#8861] [Question] [answered]
When using Pydantic for serialization and deserialization, what causes c3 and c4 not be equal? The following example fails `assert  c3 == c4`. Instead, the list `c4.data` contains `ThingA` instead of `ThingB`.

```
import json
from typing import List
from pydantic import BaseModel

class ThingA(BaseModel):
    messages: List[List[str]] | None = None
    
class ThingB(BaseModel):
    messages: List[tuple[str, str, str]] | None = None

class ThingCollection(BaseModel):
    data: List[ThingA| ThingB] | None = None

if __name__ == "__main__":
    a = ThingA(messages=[["a","a","a"],["a","a","a"],["a","a","a"]])
    c1 = ThingCollection(
        data=[a,a,a]
    )
    data_a = c1.model_dump_json()
    data_a = json.loads(data_a)
    c2 = ThingCollection(**data_a)

    assert  c1 == c2

    b = ThingB(messages=[("b","b","b"),("b","b","b"),("b","b","b")])

    c3 = ThingCollection(
        data=[b,b,b]
    )
    data_b = c3.model_dump_json()
    data_b = json.loads(data_b)
    c4 = ThingCollection(**data_b)

    assert  c3 == c4
```

**Comment by GDYendell** [comment, marked as answer]
This is because the serialised form of `ThingA` and `ThingB` are completely equivalent after the round trip. The `tuple` becomes `list`, so in both cases you get `list[list[str]]`, which can be validated successfully as either `ThingA` or `ThingB`, whichever one it tries first. 

If you change the order of your union to `ThingB | ThingA`, you will notice that you get the opposite problem - `a` becomes `ThingB` after the round trip, and (if you comment out the first assert so that it runs), `b` stays as `ThingB` after the round trip.

Members of a union are validated in the order they are defined and the first successful attempt is used. If you need to distinguish between these two types when validating, you will need to make a [discriminated union](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions). By adding a literal field to the model that gets serialised with the data, pydantic can then be use that field to discriminate the correct type during validation.

**Comment by jhb123** [comment]
Thank you for that explanation 😁
